### ISSUE TITLE
Is there a way to i18n routes?

### ISSUE THREAD

**Issue by acstll** [sveltejs/sapper#78]
If I wanted to have a multilanguage app, and have a route like `/en/bananas` and then its translation at `/ca/platans`, rendering the same component but in a different language… would there be a way to achieve this without duplicating all component files?

**Comment by Crisfole** [comment]
@acstll You can use a slug in a folder name:

 - [lang]
 - - - en.html
 - - - de.html
 - - - es.html

I'm not positive on this though

**Comment by acstll** [comment]
Nice :)

And how about translating also the routes themselves?, like where "foo" would be a `slug`

- `/en/artists/foo`
- `/de/kunstler/foo`

I would just want to reuse the same component for [slug] but not have it twice, in `artists/` and `kunstler/`… any ideas?

**Comment by Crisfole** [comment]
Ooh, that's tricky...I love the idea, but at this point I think I have to defer to @Rich-Harris.

At the moment I'm guessing you should just use Klingon ;) it should be culturally neutral enough...

**Comment by Conduitry** [comment]
Not the most graceful solution, but I'd imagine you could write some middleware that comes before Sapper to normalize the URLs.

**Comment by Crisfole** [comment]
@Conduitry that's exactly the right solution, actually. Not sure why I didn't think of it. Esp. if your first segment is the language segment, you'd just go through each path segment passing it through the translation dictionary for that language if the key is present:

```
// en.json
{
  "artists": "artista",
  "foo": "fu"
}
// de.json
{
  "kunstler": "artista",
  "foo": "fö"
}
```

Here I'm presuming you'll work in spanish on your file system.

The middleware would just replace the requested url with the canonical url, by converting each segment (if possible). Sapper would see the request as if all requests were made to the canonical spanish version (not pretty code, not tested code, but gives the general idea):

```
const translations = loadTranslations(); // Implement this
function translate(reqUrl) {
  let splitPath = url.parse(reqUrl).pathname.split('/'),
       lang = splitPath[0],
       path = splitPath.slice(1); 
  path = path.map(p => translations[lang][p] || p);
  return [lang, path...].join('/');
}

app.use(function (req, res, next) {
  req.url = translate(req.url); // req.originalUrl still present if needed
  next();
});
```

**Comment by acstll** [comment]
Sounds like a very reasonable solution. I can imagine one could even use something like `path-to-regexp` to match URLs and only translate (normalize) them if it matches something…

I'll do some tests as soon as I have some free time (and let you know).

Thank you both @Crisfole @Conduitry

**Comment by aubergene** [comment]
This can now be done as

```
routes/
├── [lang]
│   └── [kind]
│       └── index.html
```

Then within `index.html` you can use `{params.lang}`, `{params.kind}` etc.

**Comment by laurentpayot** [comment]
@acstll @Crisfole @Conduitry 
I implemented the server-side req.url translation solution (and added a client-side link urls translation function). The problem is that Sapper doesn't have the translated url in its routes, so the client will always treat the translated url as an external url, and will call the server to load this "unknown" page. I had to abandon this solution and ended-up using locale subdomains (airbnb style) instead of subdirectories. It is just as good for SEO and much simpler. For Single Page Applications you shouldn't have to many (english only) directories names anyway, and the slugs are always in the right language.
